### PR TITLE
json rpc & indexer: last piece to make indexer txn response complete

### PR DESCRIPTION
**Pull request by gegaowp**
## Description 

main things are:
- balance changes and object changes refactoring on JSON RPC to unblock indexer
- indexer balance and object changes impl
- read events from PG

## Test Plan 

CI with integration test changes

### CODE REVIEW

File: crates/sui-indexer/src/balance_changes.rs [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+use sui_types::gas::GasCostSummary;
+use sui_types::gas_coin::GAS;
+use sui_types::object::{Object, Owner};
+
+use crate::errors::IndexerError;
+use crate::store::PgIndexerStore;
+
+pub async fn get_balance_changes_from_effect<P: ObjectProvider<Error = E>, E>(
+    object_provider: &P,
+    effects: &SuiTransactionEffects,
```
Comment (gegaowp) [inline crates/sui-indexer/src/balance_changes.rs:25]: a couple of reasons to put these in sui-indexre:
- the trait is defined in json rpc, which is a dependency crate indexer
- secondary one: json rpc types like SuiTransactionEffects are used in indexer

Comment (wlmyng) [reply crates/sui-indexer/src/balance_changes.rs:25]: Is this a rust thing where traits need to copied? Like we can't just import the `ObjectProvider` trait

File: crates/sui-indexer/src/balance_changes.rs [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+) -> Result<Vec<BalanceChange>, E> {
+    let gas_owner = effects.gas_object().owner;
+
+    // Only charge gas when tx fails, skip all object parsing
+    let sui_gas_cost_summary = effects.gas_cost_summary();
+    let gas_cost_summary = GasCostSummary {
```
Comment (wlmyng) [inline crates/sui-indexer/src/balance_changes.rs:31]: doesn't look like we use `sui_gas_cost_summary` or `gas_cost_summary` other than in L37-40, are you not able to do `amount: effects.get_cost_summary()` directly

File: crates/sui-indexer/src/store/pg_indexer_store.rs [hunk 3]
```diff
@@ -80,6 +87,51 @@ impl PgIndexerStore {
...
+
+    pub async fn find_sui_types_object_lt_or_eq_version(
+        &self,
+        id: &ObjectID,
+        version: &SequenceNumber,
+    ) -> Result<Option<sui_types::object::Object>, IndexerError> {
+        let pg_object = read_only!(&self.cp, |conn| {
+            objects_history::dsl::objects_history
+                .filter(objects_history::object_id.eq(id.to_string()))
+                .filter(objects_history::version.le(version.value() as i64))
```
Comment (wlmyng) [inline crates/sui-indexer/src/store/pg_indexer_store.rs:120]: what are your thoughts on lifting
```
                .filter(objects_history::object_id.eq(object_id.to_string()))
                .filter(objects_history::version.eq(version.value() as i64))
``` into its own function

File: crates/sui-indexer/src/store/pg_indexer_store.rs [hunk 4]
```diff
@@ -233,6 +285,64 @@ impl IndexerStore for PgIndexerStore {
...
+        Ok(SuiTransactionFullResponse {
+            digest: tx_digest,
+            transaction,
+            raw_transaction: tx.raw_transaction,
+            effects,
+            confirmed_local_execution: tx.confirmed_local_execution,
+            timestamp_ms: tx.timestamp_ms as u64,
+            checkpoint: tx.checkpoint_sequence_number as u64,
+            events: SuiTransactionEvents { data: events.data },
+            object_changes: Some(object_changes),
```
Comment (wlmyng) [inline crates/sui-indexer/src/store/pg_indexer_store.rs:341]: question: probably a premature optimization haha
But in the `get_transaction_with_options` endpoint, users can specify whether they want to see these changes. However in the PR I have at https://github.com/MystenLabs/sui/pull/9669 we'll always fetch the `SuiTransactionFullResponse` and then filter for what the caller wants. I think this is the case for other functions such as `get_object` in `pg_indexer_store`. Would we incur significant latency by always fetching `object_changes` and `balance_changes`? Would this matter to the end user?

Comment (wlmyng) [reply crates/sui-indexer/src/store/pg_indexer_store.rs:341]: ah hold on i may have misunderstood this

Comment (wlmyng) [reply crates/sui-indexer/src/store/pg_indexer_store.rs:341]: ah yeah I think in crates/sui-json-rpc/src/read_api.rs there is a `if opt.show_` that adds to `futures`

File: crates/sui-indexer/src/balance_changes.rs (hunk 1 above)
Comment (gegaowp) [reply crates/sui-indexer/src/balance_changes.rs]: doing it wrong, I should do that indexer indeed, but not for arc

File: crates/sui-indexer/src/store/pg_indexer_store.rs (hunk 3 above)
Comment (gegaowp) [reply crates/sui-indexer/src/store/pg_indexer_store.rs:120]: we will have some query builder to compose all kind of query fragments, here 2 lines seem fragmented to do it separately.

File: crates/sui-indexer/src/balance_changes.rs (hunk 2 above)
Comment (gegaowp) [reply crates/sui-indexer/src/balance_changes.rs]: `effects.get_cost_summary()` is the JSON RPC type, we need the `sui-core` type, so what I was doing was just converting them, added an into instead.

**Review by 666lcz — APPROVED** [review]

**Review by 666lcz — APPROVED** [review]

**Pull request merged**